Apps/Common/src/ErrorHandling/Exceptions/HealthGatewayException.cs

```diff
@@ -0,0 +1,82 @@
+﻿// -------------------------------------------------------------------------
+//  Copyright © 2019 Province of British Columbia
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// -------------------------------------------------------------------------
+namespace HealthGateway.Common.ErrorHandling.Exceptions
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Net;
+
+    /// <summary>
+    /// Exception class for Health Gateway.
+    /// </summary>
+    [SuppressMessage("Design", "CA1032:Implement standard exception constructors", Justification = "The constructors should be explicit")]
+    public class HealthGatewayException : Exception
```

Comment:
Suggest making this class abstract so it can't be instantiated directly.

I see it's currently being used in 4 places to throw some database errors, so it may be worth considering adding a DatabaseException temporarily to use instead.

These particular cases look like we're catching an exception in the delegate and storing the exception message (or a custom message), then throwing a brand new exception in the service using the message.  We lose information about the original exception doing this.  I think eventually we should do away with DbResult and let the exceptions bubble up, so the service can decide if they want to handle.  Then we wouldn't need to throw our own DatabaseException.

Also in these cases, it looks like some of the error conditions may be more appropriately thrown as a NotFoundException (part of `resourceDelegateDelegate.InsertAsync()`) or AlreadyExistsException (part of `profileDelegate.UpdateAsync()`) if we were to refactor them.

Reply:
Sure its more boilerplate but I can do that to keep the `HealthGatewayException` class from being abused.

Reply:
After discussion, we agreed that it would be preferable to define specific exceptions rather than throw uncategorized instances of the base HealthGatewayException.